lib/oli/resources/collaboration.ex

```diff
@@ -0,0 +1,286 @@
+defmodule Oli.Resources.Collaboration do
+  alias Oli.Authoring.Course
+  alias Oli.Authoring.Course.Project
+  alias Oli.Publishing
+  alias Oli.Publishing.{DeliveryResolver, Publication}
+  alias Oli.Delivery
+  alias Oli.Delivery.{DeliverySetting, Sections}
+  alias Oli.Delivery.Sections.{Section, SectionResource}
+  alias Oli.Resources
+  alias Oli.Resources.{ResourceType, Revision}
+  alias Oli.Resources.Collaboration.{CollabSpaceConfig, Post}
+  alias Oli.Repo
+
+  import Ecto.Query, warn: false
+  import Oli.Utils
+
+  # ------------------------------------------------------------
+  # Collaborative Spaces
+
+  @doc """
+  Creates a new collaborative space (resource + revision) and attach it to the page_slug
+  revision in the project.
+
+  ## Examples
+
+      iex> create_collaborative_space(%{status: :active, ...}, %Project{}, "slug", 123)
+      {:ok,
+        %{
+          cs_resource: %Resource{},
+          cs_revision: %Revision{},
+          cs_published_resource: %PublishedResource{},
+          project: %Project{},
+          publication: %Publication{},
+          page_resource: %Resource{},
+          next_page_revision: %Revision{}
+        }
+      }
+
+      iex> create_collaborative_space(%{status: :active, ...}, %Project{}, "invalid_slug", 123)
+      {:error, {:error, {:not_found}}}
+  """
+  @spec create_collaborative_space(map(), %Project{}, String.t(), integer()) ::
+          {:ok, map()} | {:error, String.t()}
+  def create_collaborative_space(attrs, %Project{} = project, page_slug, author_id) do
+    attrs =
+      Map.merge(attrs, %{
+        author_id: author_id,
+        resource_type_id: ResourceType.get_id_by_type("collabspace")
+      })
+
+    Repo.transaction(fn ->
+      with {:ok, %{resource: cs_resource, revision: cs_revision}} <-
+            Course.create_and_attach_resource(project, attrs),
```

Comment:
It seems like the collab space is being created in a new resource, then this new resource is referenced from a page revision. We don't need the new resource.  To create/enable a collab space in a page, all that should be necessary to do is to create a new revision of that page, with the collab_space_config set to `:enabled`